ls1intum/Artemis
# `Notifications`: Send an email when Artemis detects duplicated test cases during the creation of programming exercises

### Checklist
#### General

### Motivation and Context
This PR adds email support for the NotificationType DUPLICATE_TEST_CASES, i.e. if duplicate test cases (so far only relevant for programming exercises) occur (these are tests in the TEST repository with the same name (they have to be in different Classes!).
This is a critical issue that blocks the correct workflow for students and their result processing.
Therefore @krusche asked me to add email support so Instructors and Editors will be informed via email (because the webapp can be overlooked, etc.)

There are no settings for this NotificationType/Email. You can not deactivate them due to their criticality. (This is a requirement by @krusche)

I do not change any of the logic that decides when to trigger such a notification. I simply added email support for the legacy implementation. _(The majority of the LoC are due to new tests)_

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise
- TS1 (only here the email support is enabled)
- Make sure that you use your tum account (e.g. de42sos) so it is possible for you to receive and look at the email!
- You have to be an instructor of the course (admin might not work)

I highly recommend simply following the steps in the video to trigger the duplicate test case notification/email. _(It took me way too much time to figure out how to trigger this notification, so please follow the video)_
I repeat: I did not change the logic of when or why this notification is triggered, I just added the email. The goal is to check if the email is created and it is valid.
_*It is not necessary to click the last button, the notification should appear after you have added the duplicated test case. Otherwise 2 notifications/email might/will be created_
(I am using the same user in two tabs (no incognito, etc. needed for this one) One tab is also enough

https://user-images.githubusercontent.com/65814168/144247861-58c60605-6157-41fd-a747-858f5dbc9c66.mp4

Written steps:
- create a new default programming exercise
- open it as an instructor (in course management) and click on the "edit in editor" button. 
- change the "repostory"/tab to TEST.

- Look at the File Browser to your left
- Click on the test folder and add a new file: e.g. "PlagiarismTest.java"
- edit this new file:
- e.g. add the following code into it :
```
package plagiarismHope;

import org.junit.jupiter.api.*;
import static org.junit.jupiter.api.Assertions.*;

import java.lang.reflect.InvocationTargetException;
import java.text.*;
import java.util.*;

import static de.tum.in.test.api.util.ReflectionTestUtils.*;

import de.tum.in.test.api.BlacklistPath;
import de.tum.in.test.api.PathType;
import de.tum.in.test.api.StrictTimeout;
import de.tum.in.test.api.WhitelistPath;
import de.tum.in.test.api.jupiter.Public;

@Public
@WhitelistPath("target") // mainly for Artemis
@BlacklistPath("target/test-classes") // prevent access to test-related classes and resources
class PlagiarismTest {

    @Test
    @StrictTimeout(1)
    void testMergeSort() {
        fail("MergeSort does not sort correctly");
    }
}
```

- press submit.

- check notification & email

### Review Progress
#### Code Review
#### Manual Tests

### Test Coverage
I simply added a new email.html (case) to the already working, unchained notification-email pipeline.
So no change to the logic.
I actually added a couple of tests:
- I added/adjusted test for GroupNotificationFactory for the notification type DUPLICATE_TEST_CASES
- I created a new test class NotificationTargetTest.java which tests methods that are used by the MailService to extract the correct URL from the notification.target property (this class is needed to be able to generate a valid link/URL based on a notification (prior to notification emails this was not necessary on the server-side (i.e. the URL was extracted only on the client-side)
(NotificationTarget ~ 100%)
(GroupNotificationFactory > 90%)

### Screenshots
Notification
![image](https://user-images.githubusercontent.com/65814168/144242828-503c532f-73c9-4c0e-a130-71e56bb37b18.png)

Email
![image](https://user-images.githubusercontent.com/65814168/144242840-4d013f86-d7a7-41fa-b45b-36d643401ad2.png)
